app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -244,7 +246,19 @@ private void uiStateScreenTransition() {
                 break;
 
             default:
-                fragment = installFragment;
+                List<Fragment> fgmts = fm.getFragments();
+                int lastIndex = fgmts != null ? fgmts.size() - 1 : -1;
+                if(lastIndex > -1) {
+                    fragment = fgmts.get(lastIndex);
+                    if (BuildConfig.DEBUG) {
+                        Log.v(TAG, "Last fragment: " + fragment);
+                    }
+                }
+                if(!(fragment instanceof SetupEnterURLFragment)){
+                    fragment = installFragment;
+                } else {
+                    // we restore the state of the SetupEnterURLFragment if it is our last fragment
+                }

```

Comment:
There is a fair amount of logic going on here. Could you move it into a helper method with a good name so that it is easier to understand what the 'default' case is achieving?
Maybe:
```
fragment = restoreInstallSetupFragment()
...
private Fragment restoreInstallSetupFragment() { ...
```